nixos/modules/services/misc/nixos-needsreboot.nix

```diff
@@ -0,0 +1,29 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.services.nixos-needsreboot;
+in
+{
+  options = {
+    services.nixos-needsreboot = {
+      enable = lib.mkEnableOption "nixos-needsreboot, which determines if you need to reboot your NixOS machine";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [{
+      assertion = config.system.autoUpgrade.enable == true;
+      message  = "Please enable 'system.autoUpgrade'.";
+    }];
+
+    systemd.services.nixos-needsreboot = {
+      description = "Determine if you need to reboot your NixOS machine";
+      requires = [ "nixos-upgrade.service" ];
+      wantedBy = [ "multi-user.target" ];
+      serviceConfig = {
```

Comment:
Do we want to do any hardening?

Reply:
I'm not sure what you mean by hardening

Reply:
https://wiki.nixos.org/wiki/Systemd/Hardening

An example: https://github.com/NixOS/nixpkgs/blob/c1a16412b44d133c7f6a37ad993bfd7323bd9180/nixos/modules/services/hardware/kanata.nix#L117-L148

Reply:
I don't think so. It only compares a few filesystem paths and determines version from it. The only write operation it performs is to write what critical "package" was upgraded that necessitates a reboot in the `/var/run/reboot-required` file.

Reply:
That said, what do you think about hardening? Do you feel like it's needed, if yes, what do you want? I'm open to hearing it!

Reply:
Make everything but that file read only, block network access, etc.

Reply:
ack, will push again with these three fixes:
1. [fix assertion ](https://github.com/NixOS/nixpkgs/pull/303764#discussion_r1584672534)
2. [squash last two commits](https://github.com/NixOS/nixpkgs/pull/303764#pullrequestreview-2031100030)
3. [enable hardening](https://github.com/NixOS/nixpkgs/pull/303764#discussion_r1587055159)